src/bootstrap/test.rs

```diff
@@ -704,6 +704,39 @@ impl Step for RustdocUi {
     }
 }
 
+#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
+pub struct ErrorCodesCheck {
+    pub compiler: Compiler,
+}
+
+impl Step for ErrorCodesCheck {
+    type Output = ();
+    const DEFAULT: bool = true;
+    const ONLY_HOSTS: bool = true;
+
+    fn should_run(run: ShouldRun<'_>) -> ShouldRun<'_> {
+        run.path("src/tools/error-codes")
+    }
+
+    fn make_run(run: RunConfig<'_>) {
+        let compiler = run.builder.compiler(run.builder.top_stage, run.host);
+
+        run.builder.ensure(ErrorCodesCheck { compiler });
+    }
+
+    fn run(self, builder: &Builder<'_>) {
+        builder.ensure(compile::Std { compiler: self.compiler, target: self.compiler.host });
+
+        let mut cmd = builder.tool_cmd(Tool::ErrorCodesCheck);
```

Comment:
Does this really need to be done using a new tool? Rustbuild can already run `rustdoc --test` on all `.md` files in a directory using the `DocTest` step.

Either way this will be slower than the current `ErrorIndex` step because it will have to call `rustdoc --test` once for each error code rather than just once for all of them.

Reply:
You need to write a tool in any case to import the markdown files. So instead of having everything at once, at least we get a very precise error output. Also, the run is a bit longer, but nothing much bigger.